### PR TITLE
Add schema to consumer Avro messages

### PR DESCRIPTION
**Pull request by saabeilin**
https://github.com/confluentinc/confluent-kafka-python/issues/449

### CODE REVIEW

**Comment by saabeilin** [comment]
[clabot:check]

File: confluent_kafka/avro/serializer/message_serializer.py [hunk 1]
```diff
@@ -59,6 +59,13 @@ def __exit__(self, *args):
         return False
 
 
+class AvroRecord(dict):
```
[the remark this answers is not in the document]
Comment (saabeilin) [reply confluent_kafka/avro/serializer/message_serializer.py]: But the decoder returns a dict, does not it? At least the avro system tests pass for me (not the producer tests though, but it's another story)

**Comment by saabeilin** [comment]
avro system tests pass for me both for Python 3.7 and 2.7; broker and schema registry - CP5 running in Docker.
```
$ python examples/integration_test.py --conf tests/testconf.json --avro
Using confluent_kafka module version 0.11.5 (0xb0500)
Using librdkafka version 0.11.5 (0xb05ff)
Testing: ['avro']
Brokers: localhost:29092
Topic prefix: librdkafka-test
('==============================', 'Verifying AVRO', '==============================')
('==============================', 'Done', '==============================')
```

File: confluent_kafka/avro/serializer/message_serializer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rnpridgeon) [reply confluent_kafka/avro/serializer/message_serializer.py]: You are right we are not actually wrapping the Message type here but instead the decoded contents of the Message.Key and or Message.Value. That said records map to dicts, but primitives map to their python counterparts. i.e. avro float = python float.  

https://github.com/apache/avro/blob/master/lang/py/src/avro/io.py#L28-L37

This looks to be true of Fast Avro as well based on a cursory scan of the source. 

https://github.com/fastavro/fastavro/blob/c2e2c690d93de6ee26570c1b0dc6bd92ddac1724/fastavro/_read.pyx#L245-L263

**Comment by rnpridgeon** [comment]
@saabeilin It doesn't look like your tests are actually running. You should see output similar to the following: 

```
Testing: ['avro']
Brokers: localhost:29092
Topic prefix: test-topic
('==============================', 'Verifying AVRO', '==============================')
6fbc5506-f06e-4cfc-a485-0891f8f92575[0]@0: key=32.0, value=None, tstype=1, timestamp=1537401416240
2ab22cd3-7934-4f9c-aec0-7f8fd8385b14[0]@0: key=None, value=32.0, tstype=1, timestamp=1537401422490
cb07eb50-8ab6-440c-80b3-70274a3c2eef[0]@0: key={u'number': None, u'name': u'abc'}, value=None, tstype=1, timestamp=1537401428616
```

Since we can't rely on the reader returning exclusively dict objects, see earlier comment, we should instead enhance the cimpl.Message type to allow the use of `setattr()` and `getattr()`. 

In order to achieve this we need to provide the Message type with two function pointers; One for the getter and another for the setter. These pointers are then placed within the type `tp_getattro` and `tp_setattro` slots. As you can see in the link below the getter has already been set so we just need to set the setter in a similar way. 

If the I understand the C api docs correctly you should be able to set the following slot with `PyObject_GenericSetAttr`. 

https://github.com/confluentinc/confluent-kafka-python/blob/master/confluent_kafka/src/confluent_kafka.c#L571

Further down you can see where `PyObject_GenericSetAttr` already fills the `tp_getattro` slot. 

https://github.com/confluentinc/confluent-kafka-python/blob/master/confluent_kafka/src/confluent_kafka.c#L580

The C API docs mention that setting this to the generic setter will provide the message type with "normal" attribute setting capabilities. 

https://docs.python.org/2/c-api/typeobj.html#c.PyTypeObject.tp_setattro

**Comment by saabeilin** [comment]
@rnpridgeon At least I see quite some activity in the broker and registry logs. I'll have to check the tests code thoroughly.

I really do not want to modify the Message but only inject schemas into avro-decoded key and value.

I'll close this PR for now and take a second iteration; thanks for comments. We can discuss other approaches via slack, I messaged you.

**Pull request closed without merging**